contracts/two-party-pol-holder/src/msg.rs

```diff
@@ -1,40 +1,243 @@
+    /// liquid pooler address
     pub next_contract: String,
+    /// config describing the agreed upon duration of POL
     pub lockup_config: Expiration,
+    /// config describing early exit dynamics
     pub ragequit_config: RagequitConfig,
+    /// deadline for both parties to deposit their funds
     pub deposit_deadline: Expiration,
+    /// config describing the covenant dynamics
     pub covenant_config: TwoPartyPolCovenantConfig,
+    /// mapping of denoms to their splits
+    pub splits: BTreeMap<String, SplitType>,
+    /// a split for all denoms that are not covered in the
+    /// regular `splits` list
+    pub fallback_split: Option<SplitConfig>,
 }
 
 impl InstantiateMsg {
-    pub fn get_response_attributes(self) -> Vec<Attribute> {
+    pub fn get_response_attributes(&self) -> Vec<Attribute> {
+        let fallback_attr = match self.fallback_split.as_ref() {
+            Some(split) => split.get_response_attribute("fallback_split".to_string()),
+            None => Attribute::new("fallback_split".to_string(), "none".to_string()),
+        };
+        let splits_attr: Vec<Attribute> = self
+            .splits
+            .iter()
+            .map(|(denom, split_type)| match split_type {
+                SplitType::Custom(split_config) => {
+                    split_config.get_response_attribute(denom.to_string())
+                }
+            })
+            .collect();
+
         let mut attrs = vec![
-            Attribute::new("clock_addr", self.clock_address),
-            Attribute::new("pool_address", self.pool_address),
-            Attribute::new("next_contract", self.next_contract),
+            Attribute::new("clock_addr", self.clock_address.to_string()),
+            Attribute::new("pool_address", self.pool_address.to_string()),
+            Attribute::new("next_contract", self.next_contract.to_string()),
             Attribute::new("lockup_config", self.lockup_config.to_string()),
             Attribute::new("deposit_deadline", self.deposit_deadline.to_string()),
+            fallback_attr,
         ];
         attrs.extend(self.ragequit_config.get_response_attributes());
+        attrs.extend(splits_attr);
         attrs
     }
 }
 
+#[cw_serde]
+pub enum CovenantType {
+    Share,
+    Side,
+}
+
+impl CovenantType {
+    pub fn get_response_attribute(&self) -> Attribute {
+        Attribute::new(
+            "covenant_type",
+            match self {
+                CovenantType::Share => "share",
+                CovenantType::Side => "side",
+            },
+        )
+    }
+}
+
+#[cw_serde]
+pub struct DenomSplits {
+    pub explicit_splits: BTreeMap<String, SplitConfig>,
+    pub fallback_split: Option<SplitConfig>,
+}
+
+impl DenomSplits {
+    pub fn get_fallback_distribution_messages(self, available_coins: Vec<Coin>) -> Vec<CosmosMsg> {
+        available_coins
+            .iter()
+            .filter_map(|c| {
+                // explicit splits are distributed via claim/ragequit
```

Comment:
`get_fallback_distribution_messages` is for the bogus denoms, but indeed it's still susceptible to running out of gas. Do you think pagination is the way to go, or taking in a list of denom strings as an argument to this call?

Reply:
Because its a specific message for bogus tokens, we want to make sure we just cover all of them.
I think the main issue here is the not the length of the array of tokens, but rather the fact we loop over each one of them, so what you can do is just loop over 100 at a time, and if the array is empty, it means we not longer have tokens available.
Im not sure if you do it or not, but you also need to make sure you don't loop over the important denoms here.

Reply:
Yeah this is a separate method from the important denom distribution. I filter them out just to be sure here in the `filter_map`:
```rust
if self.explicit_splits.contains_key(&c.denom) {
    None
}
```

Regarding the loop, indeed, it should eventually go down to 0 available denoms and complete. My fear is that if someone was to print a gazillion tokens, then (I think?) `query_all_balances` itself may fail. I'm basing this on that the [sdk bank module all balances query](https://github.com/cosmos/cosmos-sdk/blob/0c633a59f3afe13b6385cb0d67d55e6d53332ca3/proto/cosmos/bank/v1beta1/query.proto#L143) contains a pagination param and the following comment:
```rust
// When called from another module, this query might consume a high amount of
// gas if the pagination field is incorrectly set.
```
but cw `query_all_balances` doesn't expose that pagination field.

Following on the previous point, another infeasible scenario would be someone actively printing new tokens onto the contract. Assuming `AllBalances` query returns a list of denoms ordered alphabetically (..does it?), some potentially valuable denom could get continuously frontrun by the newly received ones.

@uditvira  what do you think?

Reply:
Ben and I discussed this on a call today. Summary is:
1. We must only every query balances or distribute tokens from a whitelist
2. We can allow third parties to distribute out tokens that are not in the whitelist if they pay for gas

This should prevent a DOS attack.

Reply:
done